feat: compress and decompress objects

Dudleyneedham opened the pull request:
## fixes https://github.com/KILTprotocol/ticket/issues/286


Adding compressing and decompressing methods to each of the objects that are being sent via messages. 

- [x] Request for Attestation
- [x] Attested Claim
- [x] Attestation
- [x] Quote
- [x] CTypes
- [x] Claim

Added additional tests to check the information

## How to test:

Yarn Test

Test Suites: 19 passed, 19 of 25 total
Tests:       68 passed, 68 total


## Checklist:

- [x] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [ ] I have documented the changes (where applicable)

maudnals commented on src/ctype/CType.ts:
```diff
@@ -18,6 +18,18 @@ import { getOwner, store } from './CType.chain'
+}
+
+export function compressCType(cType: ICType): any {
+  if (cType.owner) {
```
Change request:
You can simplify/shorten this and get rid of your `if`: 
`return [compressCTypeSchema(cType), cType.owner || null, cType.hash]`

Dudleyneedham replied:
What Timo has mentioned, he wants the code to be more verbose. However, I do think this still details what is happening.

maudnals replied:
Even more concise now 👍

maudnals commented on src/quote/Quote.spec.ts:
```diff
@@ -91,7 +91,23 @@ describe('Claim', () => {
   const invalidPropertiesQuote = invalidPropertiesQuoteData
   const invalidCostQuote = invalidCostQuoteData
+  it('compress and decompress quote objects', () => {
+    const resultCompressQuote = Quote.compressQuote(validQuoteData)
```
Change request:
Same naming change request as in my other comments.

Dudleyneedham replied:
Updated

maudnals commented on src/quote/Quote.ts:
```diff
@@ -108,3 +108,81 @@ export function createAgreedQuote(
   }
 }
+
+export function compressCost(quote: IQuote): any {
```
Change request: 
You're passing the whole `quote` object to this `compressCost` function but it doesn't need it; you can pass it simply a `cost`. It will also make the API clearer!

Dudleyneedham replied:
Updated. 

Do you prefer `ICostBreakdown` or `IQuote['cost']`?

maudnals replied:
`ICostBreakdown` as you have it is good 👍

maudnals commented on src/quote/Quote.ts:
```diff
@@ -108,3 +108,81 @@ export function createAgreedQuote(
+
+export function decompressCost(quoteArray: IQuote['cost']): any {
+  const quote = quoteArray
+  return { gross: quote[0], net: quote[1], tax: quote[2] }
```
Question:
Why are not all your objects prettified/eslinted in the same way? (see function right below)

Dudleyneedham replied:
I guess it fits all on one line.

Dudleyneedham replied:
#Unsure

maudnals commented on src/quote/Quote.ts:
```diff
@@ -108,3 +108,81 @@ export function createAgreedQuote(
+
+export function compressAttesterSignedQuote(
+  attesterSignedQuote: IQuoteAttesterSigned
+): any {
```
Change request:
You're returning an array, not `any`

Dudleyneedham replied:
Updated

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+}
+
+export function compressClaimHashTree(
+  rfa: IRequestForAttestation
```
Change request:
Unify namings; mostly in the SDK `reqForAtt` is used over `rfa` (and by the way `rfa` occurences should be replaced, but that's another story!)
This comment is valid for other new `rfa` occurences in this PR :) Once again nitpicking, but naming consistency makes it easier for other devs!

Dudleyneedham replied:
Yeah, I can update it all.

maudnals replied:
I see it's updated 👍

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+}
+
+export function compressClaimContents(rfa: IRequestForAttestation): any {
+  // should go into the claim module.
```
Question:
Is this comment intended here? Or maybe you want to make this change?

Dudleyneedham replied:
I was leaving this for a place holder to ask.

Dudleyneedham replied:
Do you think this should go inside the claim class?

maudnals replied:
I see you've updated this 👍

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+  }
+}
+
+function compressLegitimaition(leg: AttestedClaim): any {
```
Change request:
Typo in function name

Dudleyneedham replied:
Thank you.

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+  return leg.compress()
+}
+
+export function decompressLegitimaition(leg: any): any {
```
Change request:
Typo in function name

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+  reqForAttest: IRequestForAttestation
+): any {
+  return [
+    compressClaimContents(reqForAttest),
```
Question (please see w/ @tjwelde):
Do we want to do data validation here? And in general for this compression/decompression feature?

maudnals replied:
Validation is covered by `fromRequest` and similar methods that call the constructor that in turns performs validation.

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+  return Object.values(rfa.claim)
+}
+
+export function decompressClaimContents(rfa: any): any {
```
Change request:
I think the name of the function is a bit misleading; the parameter is not actually `contents`.

Dudleyneedham replied:
agreed

maudnals replied:
Fixed

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+  return Object.values(rfa.claim)
+}
+
+export function decompressClaimContents(rfa: any): any {
```
Change request:
I think the name of the function is a bit misleading; the parameter is not actually `contents`.

Dudleyneedham replied:
Double comment?! Must have struck a nerve

maudnals replied:
Right, my bad - Resolved too

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+  if (!leg[0]) {
+    return []
+  }
+  return [decompressAttestedClaim(leg[0])] // need to test for multiple leg
```
Question:
Why is an array returned here?

Dudleyneedham replied:
Because there is an array of legitimation of Attested Claims. Maybe, we can discuss. 

I would like it to look cleaner. Additionally, I need to test for multiple Legitmations.

Dudleyneedham replied:
I have tested and returning an array of legitimations.

Dudleyneedham replied:
Additionally, I iterate through them and decompress them individually

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+}
+
+export function decompressClaimHashTree(
+  rfa: IRequestForAttestation[]
```
Question:
Are you sure about the type of the function's parameter?

Dudleyneedham replied:
No, I was going to figure out this.

maudnals replied:
Much clearer thanks to your new naming 👍

maudnals commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -61,6 +63,95 @@ function getHashRoot(leaves: Uint8Array[]): Uint8Array {
+  const result = {}
+
+  Object.keys(rfa).forEach(entryKey => {
+    result[entryKey] = compressNonceAndHash(Object.values(rfa[entryKey]))
```
Suggestion:
`Object.values(rfa[entryKey])` in this function and `Object.values(claimTree[entryKey])` in the previous one are doing the same things. 
Maybe implement a helper function? Once you do so, you might also want to push it further, since you'll see that both this function and the previous one are very similar. You'll decide on the fly :)

Dudleyneedham replied:
Timo was not so keen on the `compressNonceAndHash` helper function. Since we only use this once or twice I am not sure.

But we can bring this out.

maudnals replied:
No problem 🙂

Dudleyneedham commented on src/quote/Quote.ts:
```diff
@@ -108,3 +108,81 @@ export function createAgreedQuote(
+  }
+}
+
+export function compressAgreedQuote(agreedQuote: IQuoteAgreement): any {
```
I like `quoteAgreement` more. 

I will make a PR & Ticket that will change it within the quote to be more consistent across the board.

Dudleyneedham commented on src/requestforattestation/RequestForAttestation.spec.ts:
```diff
@@ -114,6 +120,23 @@ describe('RequestForAttestation', () => {
     expect(request.claimOwner.nonce).toBeUndefined()
     expect(request.claim.owner).toBeUndefined()
   })
+  it('optimsise the request for attestation object', () => {
```
Thank you.

Dudleyneedham commented on src/quote/Quote.ts:
```diff
@@ -108,3 +108,81 @@ export function createAgreedQuote(
+}
+
+export function decompressAgreedQuote(agreedQuoteArray: any): IQuoteAgreement {
+  const agreedQuote = agreedQuoteArray
```
I have no idea. But, I literally tried so many times the other day and it wasn't working.

maudnals replied:
I see that you've removed it 👍

Dudleyneedham commented on src/attestation/Attestation.spec.ts:
```diff
@@ -77,6 +80,16 @@ describe('Attestation', () => {
     expect(await attestation.verify()).toBeFalsy()
   })
 
+  it('compress attestation', () => {
```
I have updated all the tests across the board.

Dudleyneedham replied:
What do you think? 

'compresses and decompresses the {depends on the test} object'

LeonFLK replied:
yes that's the desired grammar case

Dudleyneedham commented on src/attestation/Attestation.spec.ts:
```diff
@@ -77,6 +80,16 @@ describe('Attestation', () => {
+      requestForAttestation,
+      identityAlice
+    )
+    const optimsedResult = attestation.compress()
```
I have updated everything to match.

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -21,6 +21,20 @@ import IPublicIdentity from '../types/PublicIdentity'
 
 const log = factory.getLogger('Attestation')
 
+export function compressAttestation(attestation: IAttestation): any {
```
I have updated. 

Though, I would like to go over all the types and sort them all out, as most return any. Which is garbage.

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -21,6 +21,20 @@ import IPublicIdentity from '../types/PublicIdentity'
+  return Object.values(attestation)
+}
+
+export function decompressAttestation(attestation: any): IAttestation {
```
Updated.

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -37,6 +51,11 @@ export default class Attestation implements IAttestation {
   }
 
+  public static decompress(attestation: any): Attestation {
+    const raw = decompressAttestation(attestation)
```
What do you think of the new name?

Dudleyneedham commented on src/attestedclaim/AttestedClaim.spec.ts:
```diff
@@ -72,6 +75,12 @@ describe('RequestForAttestation', () => {
   )
 
+  it('compress attested claims', () => {
+    const optimsedResult = legitimation.compress()
```
Done.

Dudleyneedham commented on src/requestforattestation/RequestForAttestation.spec.ts:
```diff
@@ -114,6 +120,23 @@ describe('RequestForAttestation', () => {
+  it('optimsise the request for attestation object', () => {
+    const request: RequestForAttestation = buildRequestForAttestation(
+      identityBob,
+      'ctype',
```
I refuse... This is workplace bullying from all my other PRs.

Dudleyneedham commented on src/requestforattestation/RequestForAttestation.spec.ts:
```diff
@@ -114,6 +120,23 @@ describe('RequestForAttestation', () => {
+      },
+      [legitimation]
+    )
+    const optimsedResult = request.compress()
```
Updated

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -21,6 +21,20 @@ import IPublicIdentity from '../types/PublicIdentity'
 
 const log = factory.getLogger('Attestation')
 
+export function compressAttestation(attestation: IAttestation): any {
```
I was going to do it afterwards. Maybe in a separate ticket, because things will change.

LeonFLK commented:
Sorting of the property keys has to be   _sorted_   out.
We need to sort our Class property keys anyway IMO

weichweich commented on src/attestation/Attestation.ts:
```diff
@@ -21,6 +21,20 @@ import IPublicIdentity from '../types/PublicIdentity'
 
 const log = factory.getLogger('Attestation')
 
+export function compressAttestation(attestation: IAttestation): any[] {
```
you could get rid of this any (and 3 others) by using `Array<IAttestation[keyof IAttestation]>`.

Dudleyneedham replied:
Nice, I didn't know what to put, but I have never used this before. I have an idea what it is doing, but maybe explain this to me would be nice of you.

weichweich replied:
`keyof IAttestation` is a type of all the keys of IAttestation. `IAttestation[keyof IAttestation]` represents the types of all values of IAttestation. I think. [Here are the docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html#keyof-and-lookup-types).

LeonFLK replied:
It's creating a union type of all the possible keys of IAttestation so this would be IAttestation["claimHash" | "cTypeHash" | "owner" | "delegationId" | "revoked"]

weichweich commented on src/attestation/Attestation.ts:
```diff
@@ -21,6 +21,20 @@ import IPublicIdentity from '../types/PublicIdentity'
+  return Object.values(attestation)
+}
+
+export function decompressAttestation(attestation: any[]): IAttestation {
```
same here ... `Array<IAttestation[keyof IAttestation]>`...

Dudleyneedham replied:
Nice

weichweich commented on src/attestation/Attestation.ts:
```diff
@@ -21,6 +21,20 @@ import IPublicIdentity from '../types/PublicIdentity'
 const log = factory.getLogger('Attestation')
 
+export function compressAttestation(attestation: IAttestation): any[] {
+  return Object.values(attestation)
```
not sure if this is a good idea. We are relying on a fixed order. Does Object.values provide a fixed oder?

those two things are not the same(?):

```js
Object.values({"b":1, "a":2})
Object.values({"a":2, "b":1})
```

Dudleyneedham replied:
Yes, this is one of the main things I have to deal with. Currently, we have proposed jsonabc, which can sort. 

So, this will be the main thing, taking unsorted JSON objects and sorting them into our current order. E.g. what is suitable for the specific type.

weichweich commented on src/attestedclaim/AttestedClaim.spec.ts:
```diff
@@ -72,6 +75,14 @@ describe('RequestForAttestation', () => {
   )
 
+  it('compresses and decompresses the attested claims object', () => {
+    const compressedLegitimation = legitimation.compress()
```
Perhaps `legitimation` could be renamed in attested claim? This confused me a bit. 😅

Dudleyneedham replied:
Ahh, well this is because they are more than one attested claim. But, perhaps it could be changed.

weichweich commented on src/attestedclaim/AttestedClaim.ts:
```diff
@@ -155,4 +174,13 @@ export default class AttestedClaim implements IAttestedClaim {
     return result
   }
+
+  public compress(): IAttestedClaim[] {
```
The return type looks wrong. 🤔

weichweich commented on src/ctype/CType.spec.ts:
```diff
@@ -53,6 +57,16 @@ describe('CType', () => {
+
+    const compressedCType = cType.compress()
+
+    expect(compressCType(cType)).toEqual(compressedCType)
```
I am not sure if this test makes sense. This boils down to:

```js
expect(compressCType(cType)).toEqual(compressCType(cType))
```

because `cType.compress()` just called `compressCType(cType)`?

Dudleyneedham replied:
Yes, the test are incomplete. 

Timo requested me to mock what the object needs to look like and what the arrays will have to look like as well. 

I haven't gotten round to it, as a lot is changing.

weichweich commented on src/ctype/CType.ts:
```diff
@@ -16,8 +16,39 @@ import ICType from '../types/CType'
 import IClaim from '../types/Claim'
 
+export function compressCTypeSchema(cTypeSchema: ICType['schema']): any[] {
+  return [
```
this looks error prone. Perhaps some thing like this:

```js
obj = {a: 1, c: 132, xyx: 125, fs: 12, rust: "<3"}
Object.keys(obj)
  .map((v, i)=>[v, obj[v]])
  .sort((a, b) => a[0].localeCompare(b[0]))
  .map(a => a[1])
```

Dudleyneedham replied:
As previously mentioned, we are going to use the jsonabc and yes it is a problem :D

weichweich commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -52,15 +54,110 @@ function generateHashTree(contents: object): object {
+
+  Object.keys(claimTree).forEach(entryKey => {
+    result[entryKey] = Object.values(claimTree[entryKey])
+  })
```
is forEach preferred over a [for loop](https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html)?

I have not tested this but this should be equivalent?

```js
for (let entryKey in claimTree) {
  ...
}
```

Dudleyneedham replied:
The current method seems to be more verbose and fits with our current theme of keeping things that way. 

For easy of use and reading.

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -21,6 +22,39 @@ import IPublicIdentity from '../types/PublicIdentity'
+ * @returns An object that has the same properties as an [[Attestation]].
+ */
+
+export function decompressAttestation(attestation: any[]): any {
```
I have tried it, but it didn't work.

Dudleyneedham commented on src/ctype/CType.ts:
```diff
@@ -10,14 +10,86 @@
+ * @returns An ordered array of a [[CType]] schema.
+ */
+
+export function compressCTypeSchema(cTypeSchema: ICType['schema']): any[] {
```
Yep it does.

tjwelde commented on src/quote/Quote.ts:
```diff
@@ -13,9 +13,50 @@
+  IQuote['timeframe']
+]
+
+export type CompressedQuoteAttesterSigned = [
```
There is a lot of repetition going on. I didn't find a perfect solution, but what do you think about this one:
```
export type CompressedQuoteAttesterSigned = [
  CompressedQuote[0],
  CompressedQuote[1],
  CompressedQuote[2],
  CompressedQuote[3],
  CompressedQuote[4],
  CompressedQuote[5],
  IQuoteAttesterSigned['attesterSignature']
]
```

Dudleyneedham replied:
Yeah, I have updated. Let me know how you feel about it. 

Though it doesn't help with the destructuring the array.

tjwelde replied:
it's a bit better. too bad, that the spread isn't working.

tjwelde commented on src/quote/Quote.ts:
```diff
@@ -108,3 +149,156 @@ export function createAgreedQuote(
+  attesterSignedQuote: IQuoteAttesterSigned
+): CompressedQuoteAttesterSigned {
+  return [
+    attesterSignedQuote.attesterAddress,
```
You can reuse `compressQuote` here

tjwelde commented on src/quote/Quote.ts:
```diff
@@ -108,3 +149,156 @@ export function createAgreedQuote(
+  quoteAgreement: IQuoteAgreement
+): CompressedQuoteAgreed {
+  return [
+    quoteAgreement.attesterAddress,
```
You can reuse `compressAttesterSignedQuote` here.

Also look at the other functions, if you can reuse more functions.

tjwelde commented on src/requestforattestation/RequestForAttestation.spec.ts:
```diff
@@ -115,6 +124,54 @@ describe('RequestForAttestation', () => {
+      [legitimationCharlie, legitimationBob]
+    )
+
+    const sortedCompressedReqForAtt: CompressedRequestForAttestation = [
```
Doesn't need to state, that it is "sorted"

Dudleyneedham replied:
Updated.

tjwelde commented on src/requestforattestation/RequestForAttestation.spec.ts:
```diff
@@ -115,6 +124,54 @@ describe('RequestForAttestation', () => {
+    )
+
+    const sortedCompressedReqForAtt: CompressedRequestForAttestation = [
+      compressClaim(reqForAtt.claim),
```
try to avoid using these helper functions

Dudleyneedham replied:
I am not sure I can.

tjwelde replied:
I _am_ sure you can ;)

Dudleyneedham replied:
Ok, I have created this monstrosity :D

tjwelde replied:
love it

tjwelde commented on src/requestforattestation/RequestForAttestation.spec.ts:
```diff
@@ -115,6 +124,54 @@ describe('RequestForAttestation', () => {
+    )
+
+    expect(
+      RequestForAttestation.decompress(decompressedRequestForAttestationObj)
```
Also test for `RequestForAttestation .compress`

Dudleyneedham replied:
Updated.

tjwelde commented on src/requestforattestation/RequestForAttestation.spec.ts:
```diff
@@ -115,6 +124,54 @@ describe('RequestForAttestation', () => {
+    expect(
+      RequestForAttestation.decompress(decompressedRequestForAttestationObj)
+    ).toEqual(reqForAtt)
+  })
```
add negative tests

Dudleyneedham replied:
updated.

tjwelde replied:
Please rethink the negative tests

Dudleyneedham replied:
I think this is better. I have added checks. overall

tjwelde commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -52,15 +77,167 @@ function generateHashTree(contents: object): object {
+ */
+
+function decompressNonceAndHash(nonceHash: CompressedNonceHash): NonceHash {
+  // Need to fix
```
What needs to be fixed?

Dudleyneedham replied:
I left an `any` and we have cleared it already.

Dudleyneedham commented on src/attestation/Attestation.spec.ts:
```diff
@@ -71,12 +74,44 @@ describe('Attestation', () => {
+      identityAlice
+    )
+
+    const sortedCompressedAttestation: CompressedAttestation = [
```
I have updated.

Dudleyneedham commented on src/attestation/Attestation.spec.ts:
```diff
@@ -71,12 +74,44 @@ describe('Attestation', () => {
+      sortedCompressedAttestation
+    )
+
+    expect(Attestation.decompress(decompressedAttestationObj)).toEqual(
```
Updated.

Dudleyneedham commented on src/attestation/Attestation.spec.ts:
```diff
@@ -71,12 +74,44 @@ describe('Attestation', () => {
+      sortedCompressedAttestation
+    )
+
+    expect(Attestation.decompress(decompressedAttestationObj)).toEqual(
```
Added the test. 

What do you think?

tjwelde replied:
That's good

Dudleyneedham commented on src/attestation/Attestation.spec.ts:
```diff
@@ -71,12 +74,44 @@ describe('Attestation', () => {
+    expect(Attestation.decompress(decompressedAttestationObj)).toEqual(
+      attestation
+    )
+  })
```
I am not so keen on what I have done. But, it works.

tjwelde replied:
The idea of the negative tests would be, to check if your code breaks on wrong input. In this case, you could e.g. change the order in the compressed attestation and expect it to not work. If it works, then you might have to add some error handling to your functions.

Dudleyneedham replied:
Ok, I have added checks. Not sure if you will like them.

tjwelde replied:
I like them ;)

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -186,6 +234,26 @@ export default class Attestation implements IAttestation {
+   * @returns A new [[Attestation]] object.
+   */
+
+  public static decompress(attestation: IAttestation): Attestation {
```
Updated

Dudleyneedham commented on src/attestedclaim/AttestedClaim.spec.ts:
```diff
@@ -72,6 +80,44 @@ describe('RequestForAttestation', () => {
+  it('compresses and decompresses the attested claims object', () => {
+    const sortedCompressedLegitimation: CompressedAttestedClaim = [
+      [
+        compressClaim(legitimation.request.claim),
```
Again, I had some issues.

I have bloated the file and as we are generating new `nonce` and `hash`

tjwelde replied:
The original legitimation doesn't need to be expanded like that, just build the `compressedLegitimation` out of the original legitimation without using helper functions

Dudleyneedham commented on src/attestedclaim/AttestedClaim.spec.ts:
```diff
@@ -72,6 +80,44 @@ describe('RequestForAttestation', () => {
+      sortedCompressedLegitimation
+    )
+
+    expect(AttestedClaim.decompress(decompressedAttestedClaimObj)).toEqual(
```
Still having issues with this.

As I am creating a circle

tjwelde replied:
this will be solved, when you are creating the compressedAttestedClaim from the properties of the `legitimation` variable

Dudleyneedham commented on src/attestedclaim/AttestedClaim.spec.ts:
```diff
@@ -72,6 +80,44 @@ describe('RequestForAttestation', () => {
+    expect(AttestedClaim.decompress(decompressedAttestedClaimObj)).toEqual(
+      legitimation
+    )
+  })
```
Updated.

tjwelde replied:
Please rethink the negative tests

Dudleyneedham commented on src/attestedclaim/AttestedClaim.ts:
```diff
@@ -155,4 +201,24 @@ export default class AttestedClaim implements IAttestedClaim {
+   * @returns A new [[AttestedClaim]] object.
+   */
+
+  public static decompress(attestedClaim: IAttestedClaim): AttestedClaim {
```
Updated.

Dudleyneedham commented on src/claim/Claim.spec.ts:
```diff
@@ -37,4 +42,20 @@ describe('Claim', () => {
     expect(Claim.fromClaim(claimObj, testCType.schema)).toEqual(claim)
   })
+
+  it('compresses and decompresses the CType object', () => {
```
Updated.

Dudleyneedham commented on src/claim/Claim.spec.ts:
```diff
@@ -37,4 +42,20 @@ describe('Claim', () => {
+    const sortedCompressedCType: CompressedCType = [
+      testCType.hash,
+      testCType.owner,
+      compressCTypeSchema(testCType.schema),
```
Updated.

Dudleyneedham commented on src/claim/Claim.spec.ts:
```diff
@@ -37,4 +42,20 @@ describe('Claim', () => {
+
+    const decompressedCTypeObj = decompressCType(sortedCompressedCType)
+
+    expect(CType.decompress(decompressedCTypeObj)).toEqual(testCType)
```
updated.

Dudleyneedham commented on src/claim/Claim.spec.ts:
```diff
@@ -37,4 +42,20 @@ describe('Claim', () => {
+    const decompressedCTypeObj = decompressCType(sortedCompressedCType)
+
+    expect(CType.decompress(decompressedCTypeObj)).toEqual(testCType)
+  })
```
updated.

tjwelde replied:
Please rethink the negative tests. Maybe ask albi or William for help here

Dudleyneedham commented on src/ctype/CType.spec.ts:
```diff
@@ -53,6 +57,16 @@ describe('CType', () => {
+  it('compresses and decompresses the ctype object', () => {
+    const cType = CType.fromCType(fromCTypeModel)
+
+    const compressedCType = cType.compress()
```
Added 💯

Dudleyneedham commented on src/ctype/CType.spec.ts:
```diff
@@ -53,6 +57,16 @@ describe('CType', () => {
+    const compressedCType = cType.compress()
+
+    expect(compressCType(cType)).toEqual(compressedCType)
+    expect(compressCTypeSchema(cType.schema)).toEqual(compressedCType[2])
```
Updated

Dudleyneedham commented on src/ctype/CType.spec.ts:
```diff
@@ -53,6 +57,16 @@ describe('CType', () => {
+
+    expect(compressCType(cType)).toEqual(compressedCType)
+    expect(compressCTypeSchema(cType.schema)).toEqual(compressedCType[2])
+    expect(decompressCType(compressedCType)).toEqual(cType)
```
Updated

Dudleyneedham commented on src/ctype/CType.ts:
```diff
@@ -58,4 +140,24 @@ export default class CType implements ICType {
+   * @returns A new [[CType]] object.
+   */
+
+  public static decompress(cType: ICType): CType {
```
updated.

Dudleyneedham commented on src/quote/Quote.spec.ts:
```diff
@@ -91,7 +91,50 @@ describe('Claim', () => {
+      Quote.decompressQuoteAgreement(sortedCompressedResultQuoteAgreement)
+    ).toEqual(quoteBothAgreed)
+    // Need to add test
+  })
```
Updated

tjwelde replied:
Please rethink the negative tests

Dudleyneedham commented on src/quote/Quote.spec.ts:
```diff
@@ -91,7 +91,50 @@ describe('Claim', () => {
+      quoteBothAgreed.claimerSignature,
+      quoteBothAgreed.rootHash,
+    ]
+    expect(Quote.decompressQuote(sortedCompressedQuote)).toEqual(validQuoteData)
```
Updated.

tjwelde commented on src/quote/Quote.spec.ts:
```diff
@@ -91,7 +91,50 @@ describe('Claim', () => {
+      validQuoteData.currency,
+      validQuoteData.termsAndConditions,
+      validQuoteData.timeframe,
+      validAttesterSignedQuote.attesterSignature,
```
As we talked about in person, it is fine to let it be like that for now, since the effort doesn't stand in relation to the outcome

weichweich commented:
fixes KILTprotocol/ticket/issues/389

Dudleyneedham commented on src/attestation/Attestation.spec.ts:
```diff
@@ -94,4 +98,58 @@ describe('Attestation', () => {
+    ]
+    compressedAttestation.pop()
+    delete attestation.claimHash
+
```
With the Negative test, I have added the lines 138, 139 to change the structure of the data.

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -15,12 +15,71 @@ import IRequestForAttestation from '../types/RequestForAttestation'
 
 const log = factory.getLogger('Attestation')
 
+function attestationErrorCheck(attestation: IAttestation): void {
```
I took the check from inside the `Attestation` constructor and moved it as a separate function. 

As I used it to check inside the `compressAttestation` and the constructor.

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -15,12 +15,71 @@ import IRequestForAttestation from '../types/RequestForAttestation'
+export function decompressAttestation(
+  attestation: CompressedAttestation
+): IAttestation {
+  if (!Array.isArray(attestation) || attestation.length !== 5) {
```
I added a simple check to see if the data passed is an `Array` and meets the required length of the `CompressedAttestation`

Dudleyneedham commented on src/attestedclaim/AttestedClaim.spec.ts:
```diff
@@ -71,6 +75,35 @@ describe('RequestForAttestation', () => {
     {},
     []
   )
+  const compressedLegitimation: CompressedAttestedClaim = [
```
Removed the helper functions and separated the `AttestedClaims`.

Dudleyneedham commented on src/attestedclaim/AttestedClaim.spec.ts:
```diff
@@ -99,4 +132,37 @@ describe('RequestForAttestation', () => {
+
+  it('Negative test for compresses and decompresses the attested claims object', () => {
+    compressedLegitimation.pop()
+    delete legitimation.attestation
```
I have changed the data structure of the `compressedLegitimation` and `legitimation` so, I can see if they throw and error if incorrect.

Dudleyneedham commented on src/attestedclaim/AttestedClaim.ts:
```diff
@@ -8,12 +8,71 @@
 import IAttestation from '../types/Attestation'
 import IRequestForAttestation from '../types/RequestForAttestation'
 
+function attestedClaimErrorCheck(attestedClaim: IAttestedClaim): void {
```
I took the check from inside the `AttestedClaim` constructor and moved it as a separate function.

As I used it to check inside the `compressAttestedClaim` and the constructor.

Dudleyneedham commented on src/attestedclaim/AttestedClaim.ts:
```diff
@@ -8,12 +8,71 @@
+export function decompressAttestedClaim(
+  attestedClaim: CompressedAttestedClaim
+): IAttestedClaim {
+  if (!Array.isArray(attestedClaim) || attestedClaim.length !== 2) {
```
Added a simple check

Dudleyneedham commented on src/claim/Claim.ts:
```diff
@@ -12,11 +12,60 @@
+ *
+ * @returns An object that has the same properties as the [[Claim]].
+ */
+export function decompressClaim(claim: CompressedClaim): IClaim {
```
Added a simple check

Dudleyneedham commented on src/ctype/CType.ts:
```diff
@@ -10,14 +10,121 @@
+export function decompressCTypeSchema(
+  cTypeSchema: CompressedCTypeSchema
+): ICType['schema'] {
+  if (!Array.isArray(cTypeSchema) || cTypeSchema.length !== 4) {
```
Added a simple check

Dudleyneedham commented on src/ctype/CType.ts:
```diff
@@ -10,14 +10,121 @@
+ */
+
+export function decompressCType(cType: CompressedCType): ICType {
+  if (!Array.isArray(cType) || cType.length !== 3) {
```
Added a simple check

Dudleyneedham commented on src/quote/Quote.ts:
```diff
@@ -108,3 +117,269 @@ export function createAgreedQuote(
+ */
+
+export function decompressCost(cost: CompressedCostBreakdown): ICostBreakdown {
+  if (!Array.isArray(cost) || cost.length !== 3) {
```
Added a simple check

Dudleyneedham commented on src/quote/Quote.ts:
```diff
@@ -108,3 +117,269 @@ export function createAgreedQuote(
+ */
+
+export function decompressQuote(quote: CompressedQuote): IQuote {
+  if (!Array.isArray(quote) || quote.length !== 6) {
```
Added a simple check

Dudleyneedham commented on src/quote/Quote.ts:
```diff
@@ -108,3 +117,269 @@ export function createAgreedQuote(
+export function decompressAttesterSignedQuote(
+  attesterSignedQuote: CompressedQuoteAttesterSigned
+): IQuoteAttesterSigned {
+  if (!Array.isArray(attesterSignedQuote) || attesterSignedQuote.length !== 7) {
```
Added a simple check

Dudleyneedham commented on src/quote/Quote.ts:
```diff
@@ -108,3 +117,269 @@ export function createAgreedQuote(
+  quoteAgreement: CompressedQuoteAgreed
+): IQuoteAgreement {
+  if (!Array.isArray(quoteAgreement) || quoteAgreement.length !== 9) {
+    throw new Error(
```
Added a simple check

Dudleyneedham commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -52,15 +93,172 @@ function generateHashTree(contents: object): object {
+export function decompressRequestForAttestation(
+  reqForAtt: CompressedRequestForAttestation
+): IRequestForAttestation {
+  if (!Array.isArray(reqForAtt) || reqForAtt.length !== 8) {
```
Added a simple check

Dudleyneedham commented on src/requestforattestation/RequestForAttestation.ts:
```diff
@@ -22,13 +22,54 @@ import {
+import IAttestedClaim, { CompressedAttestedClaim } from '../types/AttestedClaim'
+
+function requestForAttestationErrorCheck(
+  requestForAttestation: IRequestForAttestation
```
Moved the check from within the constructor to a separate function.

Dudleyneedham commented on src/quote/Quote.ts:
```diff
@@ -108,3 +117,269 @@ export function createAgreedQuote(
     claimerSignature: signature,
   }
 }
+/**
```
I am not sure about the checks within the `Quote`.

tjwelde replied:
Yeah, the quotes produce a lot of overhead code. But let's keep it for now and refactor later

Dudleyneedham commented on src/claim/Claim.spec.ts:
```diff
@@ -33,8 +34,47 @@ describe('Claim', () => {
+
+  it('Negative test for compresses and decompresses the Claim object', () => {
+    compressedClaim.pop()
+    delete claim.owner
```
change the structure of the data

Dudleyneedham commented on src/ctype/CType.spec.ts:
```diff
@@ -85,4 +104,31 @@ describe('CType', () => {
+  it('Negative test for compresses and decompresses the ctype object', () => {
+    compressedCType.pop()
+    delete rawCType.$id
+    delete claimCtype.hash
```
changed the structure of the data

Dudleyneedham commented on src/quote/Quote.spec.ts:
```diff
@@ -125,4 +170,58 @@ describe('Claim', () => {
+    delete validAttesterSignedQuote.currency
+    compressedResultAttesterSignedQuote.pop()
+    delete quoteBothAgreed.currency
+    compressedResultQuoteAgreement.pop()
```
changed the structure of the data

Dudleyneedham commented on src/requestforattestation/RequestForAttestation.spec.ts:
```diff
@@ -115,6 +121,138 @@ describe('RequestForAttestation', () => {
+      reqForAtt
+    )
+    compressedReqForAtt.pop()
+    delete reqForAtt.claimOwner
```
changed the structure of the data

Dudleyneedham commented on src/types/Attestation.ts:
```diff
@@ -13,3 +14,11 @@ export default interface IAttestation {
   revoked: boolean
 }
+
+export type CompressedAttestation = [
```
Moved the types

Dudleyneedham commented on src/types/AttestedClaim.ts:
```diff
@@ -2,10 +2,18 @@
+  request: IRequestForAttestation
 }
+
+export type CompressedAttestedClaim = [
```
Moved the types

Dudleyneedham commented on src/types/CType.ts:
```diff
@@ -16,3 +17,16 @@ export default interface ICType {
   schema: ICTypeSchema
 }
+
+export type CompressedCTypeSchema = [
```
Moved the types

Dudleyneedham commented on src/types/Claim.ts:
```diff
@@ -6,7 +6,15 @@ import ICType from './CType'
   owner: IPublicIdentity['address']
 }
+
+export type CompressedClaimContents = object
```
Moved the types

Dudleyneedham commented on src/types/Quote.ts:
```diff
@@ -26,3 +26,40 @@ export interface IQuoteAgreement extends IQuoteAttesterSigned {
   claimerSignature: string
 }
+
+export type CompressedCostBreakdown = [
```
Moved the types

Dudleyneedham commented on src/types/RequestForAttestation.ts:
```diff
@@ -2,24 +2,48 @@
+  rootHash: Hash
 }
+
+export type CompressedNonceHash = [string, string?]
```
Moved the types

Dudleyneedham commented:
I have added checks to the decompress to check whether it is an `Array` and matches the length of those specific functions. 

I have taken the constructor check, if they had them, turned them into a function for checking both the constructor and the compress.

I took the same concept and added it to the `Quote`. 

I have changed the data structure of the negative tests, in order for them to throw their error from their type guard as well as checking if they match the `Array` length etc. 

I moved all the `type`s into the the `type` folders corresponding to the class.

tjwelde commented on src/attestation/Attestation.ts:
```diff
@@ -15,12 +15,69 @@ import IRequestForAttestation from '../types/RequestForAttestation'
 
 const log = factory.getLogger('Attestation')
 
+function attestationErrorCheck(attestation: IAttestation): void {
```
Good, that you put this into another function.

Dudleyneedham commented on src/attestation/Attestation.ts:
```diff
@@ -15,12 +15,69 @@ import IRequestForAttestation from '../types/RequestForAttestation'
+ * @returns An object that has the same properties as an [[Attestation]].
+ */
+
+export function decompressAttestation(
```
I have added them.

Dudleyneedham commented:
I have added all the changes as discussed.

Dudleyneedham commented:
Updated the names.
